v3_async_wip/v3_async_wip/iothub_mqtt_client.py

```diff
@@ -101,41 +136,78 @@ async def _enable_twin_responses(self) -> None:
+                    request_id=request_id, status=status_code, body=response_body
+                )
+            except Exception as e:
+                logger.error(
+                    "Unexpected error ({}) while translating Twin response. Dropping.".format(e)
+                )
+                # NOTE: In this situation the operation waiting for the response that we failed to
+                # receive will hang. This isn't the end of the world, since it can be cancelled,
+                # but if we really wanted to smooth this out, we could cancel the pending operation
+                # based on the request id (assuming getting the request id is not what failed).
+                # But for now, that's probably overkill, especially since this path ideally should
+                # never happen, because we would like to assume IoTHub isn't sending malformed data
+                continue
             try:
                 await self._request_ledger.match_response(response)
+            except asyncio.CancelledError:
+                # NOTE: In Python 3.7 this isn't a BaseException, so we must catch and re-raise
+                raise
             except KeyError:
                 # NOTE: This should only happen in edge cases involving cancellation of
                 # in-flight operations
-                logger.warning("Twin response (rid: {}) does not match any request")
+                logger.warning(
+                    "Twin response (rid: {}) does not match any request".format(request_id)
+                )
+            except Exception as e:
+                logger.error(
+                    "Unexpected error ({}) while matching Twin response (rid: {}). Dropping response".format(
+                        e, request_id
+                    )
+                )
 
     async def _keep_credentials_fresh(self) -> None:
         """Run indefinitely, updating MQTT credentials when new SAS Token is available"""
         logger.debug("Starting the 'keep_credentials_fresh' background task")
         while True:
             if self._sastoken_provider:
-                logger.debug("Waiting for new SAS Token to become available")
-                new_sastoken = await self._sastoken_provider.wait_for_new_sastoken()
-                logger.debug("New SAS Token available, updating MQTTClient credentials")
-                self._mqtt_client.set_credentials(self._username, str(new_sastoken))
-                # TODO: should we reconnect here? Or just wait for drop?
+                try:
+                    logger.debug("Waiting for new SAS Token to become available")
+                    new_sastoken = await self._sastoken_provider.wait_for_new_sastoken()
+                    logger.debug("New SAS Token available, updating MQTTClient credentials")
+                    self._mqtt_client.set_credentials(self._username, str(new_sastoken))
+                    # TODO: should we reconnect here? Or just wait for drop?
```

Comment:
Why not reconnect if the new credentials are available and updated?

Reply:
Complexity. I think we probably are going to do a manual reconnect here, but it doesnt' have to be part of the PR